### PR TITLE
[Demangler] Stable parent identifier in OpaqueReturnTypeParent

### PR DESCRIPTION
**Pull request by DmT021**
`OpaqueReturnTypeParent` node now references the parent with a mangled parent name, rather than a parent pointer. This makes trees obtained from different demanglers (or calls to `Demangler::demangleSymbol`) for the same symbol equal.

The faulty symbols before this change were:
```
$s4test3fooV4blahyAA1SV1fQryFQOy_Qo_AHF
$s18opaque_return_type3fooQryFQOHo
```

### CODE REVIEW

**Comment by DmT021** [comment]
@jckarter Can you take a look and run tests, please?

cc @augusto2112

**Comment by augusto2112** [comment]
@swift-ci smoke test

File: lib/Demangling/Demangler.cpp [hunk 1]
```diff
@@ -1532,34 +1532,53 @@ NodePointer Demangler::demangleExtensionContext() {
...
-      return parent;
+    if (visitedNode.hasChildren() && visitedNode.getLastChild()->getKind() ==
+                                         Node::Kind::OpaqueReturnTypeParent) {
+      return;
     }
-    visitedNode->addChild(D.createNode(Node::Kind::OpaqueReturnTypeParent,
-                                       (Node::IndexType)parent), D);
-    return parent;
+    visitedNode.addChild(D.createNode(Node::Kind::OpaqueReturnTypeParent,
+                                      StringRef(getParentID())),
```
Comment (augusto2112) [inline lib/Demangling/Demangler.cpp:1547]: Won't this `StringRef` die when the backing `std::string parentID;` is cleared from the stack when `setParentForOpaqueReturnTypeNodes` returns?

Comment (DmT021) [reply lib/Demangling/Demangler.cpp:1547]: No, it won't. There are two types of `createNode` functions:
1) `createNode(Node::Kind K, llvm::StringRef Text)` - Creates a node with `Text` copied
2) `createNodeWithAllocatedText(Node::Kind K, llvm::StringRef Text)`, `createNode(Node::Kind K, const CharVector &Text)`, `createNode(Node::Kind K, const char *Text)` - create a node with `Text` not copied.

I'm intentionally picking the first one here.

**Review by jckarter — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**